### PR TITLE
Adding fybrik module controller with taxonomy validation

### PR DESCRIPTION
**Pull request by eletonia**
This PR makes the following changes:

- adds the Fybrik Module Controller which validates Fybrik Module against the taxonomy in the Reconciler 
- adds the `Status` field to Fybrik Module CRD structure
- adds fybrik module controller unit tests to test taxonomy validation against 2 invalid and 1 valid module
- modifies manager Makefile to generate new `taxonomy.json` with all module layers
- aligns fybrik application controller unit tests with fybrik application taxonomy

Completes some tasks in #613

### CODE REVIEW

File: manager/controllers/app/fybrikmodule_controller.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+// FybrikModuleReconciler reconciles a FybrikModule object
+type FybrikModuleReconciler struct {
+	client.Client
+	Name   string
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// Reconcile reconciles FybrikModule CRD
+// It receives FybrikModule CRD and selects the appropriate modules that will run
```
Comment (simanadler) [inline manager/controllers/app/fybrikmodule_controller.go:39]: These comments look like they were cut and pasted.  They don't appear to be a description of what this controller does.

File: manager/controllers/app/fybrikmodule_controller.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		return ctrl.Result{}, nil
+	}
+
+	observedStatus := moduleContext.Status.DeepCopy()
+	moduleVersion := moduleContext.GetGeneration()
+
+	// check if module has been validated before or if validated module is outdated
+	if string(moduleContext.Status.ValidModule) == "" || observedStatus.ValidatedGeneration != moduleVersion {
+		// do validation on moduleContext
+		err := ValidateFybrikModule(moduleContext, "/tmp/taxonomy/fybrik_module.json")
```
Comment (simanadler) [inline manager/controllers/app/fybrikmodule_controller.go:62]: Should the taxonomy location perhaps be in an environment variable rather than hard coded, or at least be in a constant?

File: manager/controllers/app/fybrikmodule_controller.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "app.fybrik.io", Kind: "FybrikModule"},
+		module.Name, allErrs)
+}
+
+// SetupWithManager registers Module controller
+func (r *FybrikModuleReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// 'UpdateFunc' and 'CreateFunc' used to judge if the event came from within the blueprint's namespace.
```
Comment (simanadler) [inline manager/controllers/app/fybrikmodule_controller.go:121]: Comment is incorrect.  Not blueprint's namespace.

File: manager/controllers/app/fybrikmodule_controller.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	// If that is true, the event will be processed by the reconciler.
+	// If it's not then it is a rogue event created by someone outside of the control plane.
+
+	moduleNamespace := utils.GetSystemNamespace()
+
+	p := predicate.Funcs{
```
Comment (simanadler) [inline manager/controllers/app/fybrikmodule_controller.go:127]: @shlomitk1 Am I correct that we don't need this, because you addressed this through RBAC configurations?

File: manager/controllers/app/fybrikmodule_controller.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		Log:    ctrl.Log.WithName("controllers").WithName(name),
+		Scheme: mgr.GetScheme(),
+	}
+}
+
+func ValidateFybrikModule(module *api.FybrikModule, taxonomyFile string) error {
+	var allErrs []*field.Error
+
+	// Convert Fybrik module Go struct to JSON
+	applicationJSON, err := json.Marshal(module)
```
Comment (simanadler) [inline manager/controllers/app/fybrikmodule_controller.go:101]: Maybe moduleJSON instead of applicationJSON

**Review by simanadler — COMMENTED** [review]
See some minor comments

File: manager/controllers/app/fybrikmodule_controller.go (hunk 4 above)
Comment (shlomitk1) [reply manager/controllers/app/fybrikmodule_controller.go:127]: Right, we don't need this.

File: manager/apis/app/v1alpha1/fybrikmodule_types.go [hunk 6]
```diff
@@ -223,6 +224,21 @@ type ChartSpec struct {
 	Values map[string]string `json:"values,omitempty"`
 }
 
+// FybrikModuleStatus defines the observed state of FybrikModule.
+type FybrikModuleStatus struct {
+	// ErrorMessage indicates that an error has happened during the reconcile, unrelated to a specific asset
```
Comment (shlomitk1) [inline manager/apis/app/v1alpha1/fybrikmodule_types.go:229]: Wrong comment, it's an error message received during the module validation.

File: manager/controllers/app/fybrikmodule_controller.go (hunk 1 above)
Comment (eletonia) [reply manager/controllers/app/fybrikmodule_controller.go]: Modified to be specific to Fybrik Module

File: manager/controllers/app/fybrikmodule_controller.go (hunk 2 above)
Comment (eletonia) [reply manager/controllers/app/fybrikmodule_controller.go]: I made it a const variable

File: manager/controllers/app/fybrikmodule_controller.go (hunk 3 above)
Comment (eletonia) [reply manager/controllers/app/fybrikmodule_controller.go]: Removed comment

File: manager/controllers/app/fybrikmodule_controller.go (hunk 4 above)
Comment (eletonia) [reply manager/controllers/app/fybrikmodule_controller.go]: Removed `p`

File: manager/controllers/app/fybrikmodule_controller.go (hunk 5 above)
Comment (eletonia) [reply manager/controllers/app/fybrikmodule_controller.go]: Fixed

File: manager/apis/app/v1alpha1/fybrikmodule_types.go (hunk 6 above)
Comment (eletonia) [reply manager/apis/app/v1alpha1/fybrikmodule_types.go]: Fixed

File: manager/controllers/app/fybrikmodule_controller.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+	observedStatus := moduleContext.Status.DeepCopy()
+	moduleVersion := moduleContext.GetGeneration()
+
+	// check if module has been validated before or if validated module is outdated
+	if string(moduleContext.Status.ValidModule) == "" || observedStatus.ValidatedGeneration != moduleVersion {
+		// do validation on moduleContext
+		const moduleTaxonomyFile = "/tmp/taxonomy/fybrik_module.json"
+		err := ValidateFybrikModule(moduleContext, moduleTaxonomyFile)
+		log.V(0).Info("Reconciler validating Fybrik module")
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikmodule_controller.go:59]: Remove this printout

File: manager/controllers/app/fybrikmodule_controller_unit_test.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+	// Set the logger to development mode for verbose logs.
+	logf.SetLogger(zap.New(zap.UseDevMode(true)))
+
+	namespaced := types.NamespacedName{
+		Name:      "module-with-interface-errors",
+		Namespace: "fybrik-system",
+	}
+	filename := "../../testdata/unittests/fybrikmodule-interfaceErrors.yaml"
+	fybrikApp := &app.FybrikModule{}
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikmodule_controller_unit_test.go:49]: change to fybrikModule (all over the file)

File: manager/controllers/app/fybrikmodule_controller.go (hunk 7 above)
Comment (eletonia) [reply manager/controllers/app/fybrikmodule_controller.go]: Done

File: manager/controllers/app/fybrikmodule_controller_unit_test.go (hunk 8 above)
Comment (eletonia) [reply manager/controllers/app/fybrikmodule_controller_unit_test.go]: Done

File: manager/controllers/app/fybrikmodule_controller.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+		return ctrl.Result{}, nil
+	}
+
+	observedStatus := moduleContext.Status.DeepCopy()
+	moduleVersion := moduleContext.GetGeneration()
+
+	// check if module has been validated before or if validated module is outdated
+	if string(moduleContext.Status.ValidModule) == "" || observedStatus.ValidatedGeneration != moduleVersion {
+		// do validation on moduleContext
+		const moduleTaxonomyFile = "/tmp/taxonomy/fybrik_module.json"
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikmodule_controller.go:57]: Define const at the beginning of the file, similar to type, instead of defining it in the middle of a function.

Comment (eletonia) [reply manager/controllers/app/fybrikmodule_controller.go]: Fixed

**Review by shlomitk1 — APPROVED** [review]

File: config/taxonomy/example/module/actions.yaml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 definitions:
   Action:
-    oneOf:
+    anyOf:
```
Comment (roee88) [inline config/taxonomy/example/module/actions.yaml:3]: Why the change? I believe that it must be oneOf.

File: manager/apis/app/v1alpha1/fybrikmodule_types.go [hunk 11]
```diff
@@ -223,6 +224,21 @@ type ChartSpec struct {
...
+	// +optional
+	ErrorMessage string `json:"errorMessage,omitempty"`
+
+	// ValidatedGeneration is the version of the FybrikModule that has been validated with the taxonomy defined.
+	// +optional
+	ValidatedGeneration int64 `json:"validatedGeneration,omitempty"`
+
+	// ValidModule indicates whether the FybrikModule is valid given the defined taxonomy
+	// +optional
+	ValidModule corev1.ConditionStatus `json:"validModule,omitempty"`
```
Comment (roee88) [inline manager/apis/app/v1alpha1/fybrikmodule_types.go:239]: @shlomitk1 is that aligned with how we usually handle readiness and errors? Should `errorMessage` be just `error`? Should `validModule` be `ready`? Should conditions be used instead? (the use of `corev1.ConditionStatus` is not within a condition).

File: manager/apis/app/v1alpha1/fybrikmodule_types.go [hunk 12]
```diff
@@ -236,7 +252,8 @@ type FybrikModule struct {
 	metav1.ObjectMeta `json:"metadata,omitempty"`
 
 	// +required
-	Spec FybrikModuleSpec `json:"spec"`
+	Spec   FybrikModuleSpec   `json:"spec"`
+	Status FybrikModuleStatus `json:"status,omitempty"`
 }
```
Comment (roee88) [inline manager/apis/app/v1alpha1/fybrikmodule_types.go:257]: Missing `+kubebuilder:subresource:status` below ``+kubebuilder:object:root=true`? Copy the annotations from one of the other manager types.

File: manager/apis/app/v1alpha1/fybrikmodule_types.go (hunk 11 above)
Comment (shlomitk1) [reply manager/apis/app/v1alpha1/fybrikmodule_types.go:239]: We could reuse the following structure that is currently defined in fybrikapplication_types.go
```
type Condition struct {
	// Type of the condition
	Type ConditionType `json:"type"`
	// Status of the condition: true or false
	Status corev1.ConditionStatus `json:"status"`
	// Message contains the details of the current condition
	// +optional
	Message string `json:"message,omitempty"`
}
```
We could even leverage the same types of Deny, Ready and Error (Deny == invalid, Ready == valid, Error == could not validate).

File: config/taxonomy/example/module/actions.yaml (hunk 10 above)
Comment (eletonia) [reply config/taxonomy/example/module/actions.yaml:3]: @roee88 `oneOf` was throwing this error for a [valid Fybrik module](https://github.com/fybrik/fybrik/blob/master/manager/testdata/unittests/implicit-copy-batch-module-csv.yaml):
``` 
INFO    test-module-controller  Fybrik module validation failed FybrikModule.app.fybrik.io "implicit-copy-batch-s3" is invalid: [spec.capabilities.0.actions.0: Invalid value: map[string]interface {}{"name":"RedactAction"}: Must validate one and only one schema (oneOf), spec.capabilities.0.actions.0: Invalid value: map[string]interface {}{"name":"RedactAction"}: RedactAction is required, spec.capabilities.0.actions.1: Invalid value: map[string]interface {}{"name":"RemoveAction"}: Must validate one and only one schema (oneOf), spec.capabilities.0.actions.1: Invalid value: map[string]interface {}{"name":"RemoveAction"}: RemoveAction is required]  {"FybrikModule": "fybrik-system/implicit-copy-batch-s3"}
```

File: manager/apis/app/v1alpha1/fybrikmodule_types.go (hunk 11 above)
Comment (eletonia) [reply manager/apis/app/v1alpha1/fybrikmodule_types.go]: @roee88 `ErrorMessage` and `ValidModule` are consistent with the `Status` fields we defined in [FybrikApplication](https://github.com/fybrik/fybrik/blob/895cc922d74d10c02f229caa291e0d41918d13f7/manager/apis/app/v1alpha1/fybrikapplication_types.go#L182)

File: manager/apis/app/v1alpha1/fybrikmodule_types.go (hunk 12 above)
Comment (eletonia) [reply manager/apis/app/v1alpha1/fybrikmodule_types.go:257]: Added

**Review by ronenkat — APPROVED** [review]
